README.md
# RARBG
[![Gem Version](https://img.shields.io/gem/v/rarbg.svg?colorB=brightgreen)](https://rubygems.org/gems/rarbg)
[![Build Status](https://travis-ci.org/epistrephein/rarbg.svg?branch=master)](https://travis-ci.org/epistrephein/rarbg)
[![Dependencies](https://badges.depfu.com/badges/32877a5a58ad7eb3f5db321d85a23c1b/overview.svg)](https://depfu.com/github/epistrephein/rarbg?project_id=5845)
[![Maintainability](https://img.shields.io/codeclimate/maintainability/epistrephein/rarbg)](https://codeclimate.com/github/epistrephein/rarbg/maintainability)
[![Test Coverage](https://img.shields.io/codeclimate/coverage/epistrephein/rarbg)](https://codeclimate.com/github/epistrephein/rarbg/test_coverage)
[![Inline Docs](https://inch-ci.org/github/epistrephein/rarbg.svg?branch=master)](https://inch-ci.org/github/epistrephein/rarbg)

Ruby client for the RARBG Torrent API.

## Installation

Install as a gem

```shell
$ gem install rarbg
```

Or add it to your Gemfile and execute `bundle install`

```ruby
gem 'rarbg', '~> 2.0'
```

## Usage

This gem wraps all API methods available from [RARBG TorrentAPI](https://torrentapi.org/apidocs_v2.txt?&app_id=rarbg-rubygem).

An authentication token is automatically generated on the first request, stored with a timestamp and renewed every 800 seconds.

Rate limit (1req/2s) is automatically enforced.

Check out the [full documentation](https://epistrephein.github.io/rarbg) for details on all methods and parameters.

#### Getting started

Require the gem and initialize a new `RARBG::API` object.

```ruby
require 'rarbg'

rarbg = RARBG::API.new
```

#### Methods

Use [`list`](https://epistrephein.github.io/rarbg/RARBG/API.html#list-instance_method) to list torrents.
Additional parameters are passed as keyword arguments and can be mixed together.

All successful method calls return an array of hashes (or an empty array for no results).

```ruby
# List last 100 torrents.
rarbg.list(limit: 100)

# List torrents with at least 20 seeders, sorted by seeders.
rarbg.list(min_seeders: 20, sort: :seeders)

# List torrents with extended json infos.
rarbg.list(format: :json_extended)
```

Use [`search`](https://epistrephein.github.io/rarbg/RARBG/API.html#search-instance_method) to search torrents.
One search type parameter among `string`, `imdb`, `themoviedb` and `tvdb` is required.

```ruby
# Search torrents using literal string query.
rarbg.search(string: 'Force Awakens')

# Search by IMDB id, in `Movies/x264/1080` and `Movies/x264/720`.
rarbg.search(imdb: 'tt2488496', category: [44, 45])

# Search unranked torrents by TheMovieDB id, sorted by last.
rarbg.search(themoviedb: 140607, ranked: false, sort: :last)
```

These methods are also available from the top module namespace for convenience.

```ruby
RARBG.list(sort: :leechers, min_leechers: 10)

RARBG.search(string: 'Star Wars', category: [48])
```

A list of name/id pairs for each category is available for quick lookup.

```ruby
RARBG::CATEGORIES
# => { "Movies/XVID"     => 14,
#      "Movies/XVID/720" => 48,
#      "Movies/x264"     => 17,
#      ...
```

#### Errors

API endpoint errors will raise a `RARBG::APIError` exception with the API error message.

```ruby
rarbg.list(sort: :name)
# => RARBG::APIError: Invalid sort

rarbg.search(string: 'Star Wars', min_seeders: 'notanumber')
# => RARBG::APIError: Invalid value for min_seeders

rarbg.search(imdb: 'tt0121765')
# => RARBG::APIError: Service unavailable (503)
```

Parameter validation errors on client side will raise an `ArgumentError`.

```ruby
rarbg.list('a string instead of an hash')
# => ArgumentError: Expected params hash

rarbg.search(limit: 50)
# => ArgumentError: One search parameter required among: string, imdb, tvdb, themoviedb
```

Lower level connection errors will raise `Faraday::Error` subclasses exceptions.

```ruby
rarbg.search(string: 'a timeout error')
# => Faraday::ConnectionFailed: execution expired
```

## Contributing

Bug reports and pull requests are welcome on [GitHub](https://github.com/epistrephein/rarbg).

This project is intended to be a safe, welcoming space for collaboration,
and contributors are expected to adhere to the [Code of Conduct](https://github.com/epistrephein/rarbg/blob/master/CODE_OF_CONDUCT.md).

You can contribute changes by forking the project and submitting a pull request. To get started:

1. Fork the repo
2. Install the dependencies (`bin/setup`)
3. Create your feature branch (`git checkout -b my-new-feature`)
4. Commit your changes (`git commit -am 'Add some feature'`)
5. Pass the test suite (`rake spec`)
6. Push to the branch (`git push origin my-new-feature`)
7. Create a new pull request

For more information about contributing to this project, check out [CONTRIBUTING](https://github.com/epistrephein/rarbg/blob/master/CONTRIBUTING.md).

## License

This gem is released as open source under the terms of the [MIT License](https://github.com/epistrephein/rarbg/blob/master/LICENSE).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.  
The format is based on [Keep a Changelog](https://keepachangelog.com/),
and this project adheres to [Semantic Versioning](https://semver.org/).

## Unreleased
[Diff](https://github.com/epistrephein/rarbg/compare/v2.0.0...master)

#### Changed
- Update RuboCop gem and Hound version to v0.80 ([#24](https://github.com/epistrephein/rarbg/pull/24)).


## 2.0.0 - 2020-01-22
[RubyGems](https://rubygems.org/gems/rarbg/versions/2.0.0) |
[Release](https://github.com/epistrephein/rarbg/releases/tag/v2.0.0) |
[Diff](https://github.com/epistrephein/rarbg/compare/v1.4.0...v2.0.0)

#### Changed
- Update Faraday to v1.0 and drop `faraday_middleware` dependency ([#17](https://github.com/epistrephein/rarbg/pull/17)).
- Update Ruby minimum version to 2.3 ([#17](https://github.com/epistrephein/rarbg/pull/17)).
- Rename attr_reader `conn` to `connection` ([#18](https://github.com/epistrephein/rarbg/pull/18)).
- Remove unnecessary files and attributes from gemspec ([#21](https://github.com/epistrephein/rarbg/pull/21)).
- Update version constraints for Rake, following v13.0 release ([#15](https://github.com/epistrephein/rarbg/pull/15)).
- Update version constraints for RuboCop in order to avoid breaking changes ([#16](https://github.com/epistrephein/rarbg/pull/16)).
- Tweak and reorder documentation ([#18](https://github.com/epistrephein/rarbg/pull/18)).


## 1.4.0 - 2019-06-05
[RubyGems](https://rubygems.org/gems/rarbg/versions/1.4.0) |
[Release](https://github.com/epistrephein/rarbg/releases/tag/v1.4.0) |
[Diff](https://github.com/epistrephein/rarbg/compare/v1.3.0...v1.4.0)

#### Fixed
- Fix typos and code style here and there.

#### Changed
- Bump RuboCop development dependency minimum version to `0.71` and update Hound
linter version ([#12](https://github.com/epistrephein/rarbg/pull/12)).
- Allow any Bundler version equal or greater than `1.15` ([#9](https://github.com/epistrephein/rarbg/pull/9))
but lesser than `3.0` ([#10](https://github.com/epistrephein/rarbg/pull/10)).


## 1.3.0 - 2019-02-07
[RubyGems](https://rubygems.org/gems/rarbg/versions/1.3.0) |
[Release](https://github.com/epistrephein/rarbg/releases/tag/v1.3.0) |
[Diff](https://github.com/epistrephein/rarbg/compare/v1.2.0...v1.3.0)

#### Fixed
- Return an empty array when endpoint responds with a `Can't find` error on
`imdb`/`themoviedb`/`tvdb` search ([#6](https://github.com/epistrephein/rarbg/pull/6)).
- Parse JSON only for responses with `application/json` Content-Type, preventing
error responses to raise a `ParsingError` ([#4](https://github.com/epistrephein/rarbg/pull/4)).

#### Added
- Add `token!` public method for standalone token generation
([#3](https://github.com/epistrephein/rarbg/pull/3)).

#### Changed
- Reduce Faraday connection timeout to 30 seconds.


## 1.2.0 - 2019-01-02
[RubyGems](https://rubygems.org/gems/rarbg/versions/1.2.0) |
[Release](https://github.com/epistrephein/rarbg/releases/tag/v1.2.0) |
[Diff](https://github.com/epistrephein/rarbg/compare/v1.1.1...v1.2.0)

#### Fixed
- Fixed gemspec to not include HTML documentation files.
- Fixed code style and typos.

#### Added
- Added rubocop as development dependency to enforce code style.
- Added GitHub templates for issues and pull requests.
- Added CONTRIBUTING document with contribution guidelines.

#### Changed
- Bumped all dependencies to stricter minor version constraint.
- Rate limit is now stubbed to 0.1 seconds in RSpec to speed up tests.


## 1.1.1 - 2018-10-19
[RubyGems](https://rubygems.org/gems/rarbg/versions/1.1.1) |
[Release](https://github.com/epistrephein/rarbg/releases/tag/v1.1.1) |
[Diff](https://github.com/epistrephein/rarbg/compare/v1.1.0...v1.1.1)

#### Changed
- Self-host documentation on GitHub pages.


## 1.1.0 - 2018-10-05
[RubyGems](https://rubygems.org/gems/rarbg/versions/1.1.0) |
[Release](https://github.com/epistrephein/rarbg/releases/tag/v1.1.0) |
[Diff](https://github.com/epistrephein/rarbg/compare/v1.0.1...v1.1.0)

#### Added
- `list` and `search` methods are now available from top level namespace.
- `RARBG::CATEGORIES` now returns name/id pairs of all RARBG categories.
- Faraday now uses the `APP_ID` as User agent.


## 1.0.1 - 2018-05-04
[RubyGems](https://rubygems.org/gems/rarbg/versions/1.0.1) |
[Release](https://github.com/epistrephein/rarbg/releases/tag/v1.0.1) |
[Diff](https://github.com/epistrephein/rarbg/compare/v1.0.0...v1.0.1)

#### Fixed
- Fixed code styling and some documentation errors.
- Fixed IMDb id autocorrect in `search`.

#### Added
- Faraday now logs requests to stdout if in verbose mode (`-w` or `-v`).


## 1.0.0 – 2018-04-02
[RubyGems](https://rubygems.org/gems/rarbg/versions/1.0.0) |
[Release](https://github.com/epistrephein/rarbg/releases/tag/v1.0.0) |
[Diff](https://github.com/epistrephein/rarbg/compare/v0.1.4...v1.0.0)

#### Changed
- Codebase has been completely refactored to be more resilient and error-free.
- This release is NOT backward compatible with 0.1 versions.
- Query parameters now use standard ruby syntax as expected and accept both
symbol and string keys.
- Compliance with API rate limiting (1req/2s) is now enforced automatically.
- Empty results now return an empty array instead of raising an exception.

#### Removed
- Removed `default_params`: query parameters are now passed per request.
- Removed `search_string`, `search_imdb`, `search_themoviedb`, `search_tvdb`:
search type must be specified as a keyword argument upon `search`.
- Removed `RequestError` exception: all API-related errors now raise an
`APIError` exception.


## 0.1.4 – 2017-12-22
[RubyGems](https://rubygems.org/gems/rarbg/versions/0.1.4) |
[Release](https://github.com/epistrephein/rarbg/releases/tag/v0.1.4) |
[Diff](https://github.com/epistrephein/rarbg/compare/v0.1.3...v0.1.4)

#### Fixed
- Set Faraday middleware before adapter to suppress warning.


## 0.1.3 – 2016-12-23
[RubyGems](https://rubygems.org/gems/rarbg/versions/0.1.3) |

spec/stubs.rb
# frozen_string_literal: true

# WebMock stubs for RSpec.
module Stubs
  API_ENDPOINT = RARBG::API::API_ENDPOINT

  def stub_token(token)
    stub_request(:get, API_ENDPOINT)
      .with(query: hash_including('get_token' => 'get_token'))
      .to_return(status: 200, body: { token: token }.to_json,
                 headers: { 'Content-Type' => 'application/json' })
  end

  def stub_list(token, params = {}, result = {})
    stub_request(:get, API_ENDPOINT)
      .with(query: hash_including(params.update(
        mode:  'list',
        token: token
      )))
      .to_return(status: 200, body: result.to_json,
                 headers: { 'Content-Type' => 'application/json' })
  end

  def stub_search(token, params = {}, result = {})
    stub_request(:get, API_ENDPOINT)
      .with(query: hash_including(params.update(
        mode:  'search',
        token: token
      )))
      .to_return(status: 200, body: result.to_json,
                 headers: { 'Content-Type' => 'application/json' })
  end

  def stub_error(status, error)
    stub_request(:get, API_ENDPOINT)
      .with(query: hash_including({}))
      .to_return(status: [status, error])
  end

  def stub_timeout
    stub_request(:get, API_ENDPOINT)
      .with(query: hash_including({}))
      .to_timeout
  end
end


spec/spec_helper.rb
# frozen_string_literal: true

require 'simplecov'
SimpleCov.start

require 'bundler/setup'
require 'webmock/rspec'
require 'securerandom'
require 'rarbg'

require_relative 'stubs'

RSpec.configure do |config|
  # Attach WebMock and its stubbed requests.
  WebMock.disable_net_connect!(allow_localhost: true)
  config.include(Stubs)

  # Enable flags like --only-failures and --next-failure.
  config.example_status_persistence_file_path = '.rspec_status'

  # Disable RSpec exposing methods globally on `Module` and `main`.
  config.disable_monkey_patching!

  # Enable temporarily focused examples and groups.
  config.filter_run_when_matching :focus

  # Use expect syntax.
  config.expect_with :rspec do |c|
    c.syntax = :expect
  end

  # Stub RATE_LIMIT to speed up tests.
  config.before(:example) do
    stub_const('RARBG::API::RATE_LIMIT', 0.1)
  end
end


spec/rarbg/list_spec.rb
# frozen_string_literal: true

RSpec.describe 'RARBG::API#list' do
  let(:cassette) { 'list/200' }
  let(:subject) { RARBG::API.new }
  let(:params) { {} }
  let(:list) do
    VCR.use_cassette(cassette) {
      subject.list(params)
    }
  end

  context 'when performing a list request' do
    it 'generate a token first' do
      expect { list }
        .to change { subject.token }
        .from(nil).to(String)
    end

    it 'respects rate limit' do
      list

      expect(subject.last_request - subject.token_time)
        .to be >= RARBG::API::RATE_LIMIT
    end
  end

  context 'when list request succeeds' do
    it 'returns and array of hashes' do
      expect(list).to all(be_an(Hash))
    end

    it 'returns hashes with filename and download link' do
      expect(list).to all include('filename').and include('download')
    end
  end

  context 'when list request returns no result' do
    let(:params) { { min_seeders: 1_000_000 } }

    it 'returns an empty array' do
      expect(list).to eq([])
    end
  end

  context 'when list request parameters is not an hash' do
    let(:params) { 'string' }

    it 'raises an ArgumentError exception' do
      expect { subject.list('string') }.to raise_error(
        ArgumentError, 'Expected params hash'
      )
    end
  end

  context 'when list request has invalid parameters' do
    let(:params) { { min_seeders: 'string' } }

    it 'raises a RARBG::APIError exception' do
      expect { list }.to raise_error(
        RARBG::APIError, 'Invalid value for min_seeders'
      )
    end
  end

  context 'when list request fails' do
    let(:cassette) { 'list/500' }

    it 'raises a RARBG::APIError exception' do
      expect { list }.to raise_error(
        RARBG::APIError, 'Internal Server Error (500)'
      )
    end
  end

  context 'when called from top level namespace' do
    let(:subject) { RARBG.clone }

    it 'instantiates an API object' do
      expect { list }
        .to change { subject.instance_variable_get(:@rarbg).class }
        .to(RARBG::API)
    end
  end
end


lib/rarbg/categories.rb
# frozen_string_literal: true

module RARBG
  # RARBG categories mapping.
  CATEGORIES = {
    'Movies/XVID'        => 14,
    'Movies/XVID/720'    => 48,
    'Movies/x264'        => 17,
    'Movies/x264/1080'   => 44,
    'Movies/x264/720'    => 45,
    'Movies/x264/3D'     => 47,
    'Movies/x264/4k'     => 50,
    'Movies/x265/4k'     => 51,
    'Movies/x265/4k/HDR' => 52,
    'Movies/Full BD'     => 42,
    'Movies/BD Remux'    => 46,
    'TV Episodes'        => 18,
    'TV HD Episodes'     => 41,
    'TV UHD Episodes'    => 49,
    'Music/MP3'          => 23,
    'Music/FLAC'         => 25,
    'Games/PC ISO'       => 27,
    'Games/PC RIP'       => 28,
    'Games/PS3'          => 40,
    'Games/PS4'          => 53,
    'Games/XBOX-360'     => 32,
    'Software/PC ISO'    => 33,
    'XXX'                => 4
  }.freeze
end


lib/rarbg/api.rb
# frozen_string_literal: true

require 'faraday'
require 'json'

# Main namespace for RARBG.
module RARBG
  # Default error class for the module. Wraps exceptions raised at API level.
  class APIError < StandardError; end

  # Base class for RARBG API.
  class API
    # RARBG API endpoint.
    API_ENDPOINT = 'https://torrentapi.org/pubapi_v2.php'

    # App name identifier.
    APP_ID = 'rarbg-rubygem'

    # Default token expiration time (seconds).
    TOKEN_EXPIRATION = 800

    # Default API rate limit (seconds).
    RATE_LIMIT = 2.1

    # The underlying `Faraday::Connection` object used to perform requests.
    #
    # @note For more info regarding this object refer to the
    #   {https://www.rubydoc.info/gems/faraday/Faraday Faraday documentation}.
    #
    # @return [Faraday::Connection] The Faraday connection object.
    attr_reader :connection

    # The monotonic timestamp of the last request performed.
    #   Used to comply with the endpoint rate limit based on {RATE_LIMIT}.
    #
    # @return [Float] The monotonic timestamp of the last request performed.
    attr_reader :last_request

    # The token used for authentication.
    #   This is generated and rate limited automatically when calling {#list}
    #   or {#search}, but can also be generated forcefully using {#token!}.
    #
    # @return [String] The token used for authentication.
    attr_reader :token

    # The monotonic timestamp of the token request.
    #   Used to compute the next required token request based on {TOKEN_EXPIRATION}.
    #
    # @return [Float] The monotonic timestamp of the token request.
    attr_reader :token_time

    # Initialize a new instance of `RARBG::API`.
    #
    # @example
    #   rarbg = RARBG::API.new
    def initialize
      @connection = Faraday.new(url: API_ENDPOINT) do |conn|
        conn.response :logger if $VERBOSE
        conn.adapter  Faraday.default_adapter

        conn.options.timeout      = 30
        conn.options.open_timeout = 10

        conn.headers[:user_agent] = APP_ID
        conn.params[:app_id]      = APP_ID
      end
    end

    # List torrents.
    #
    # @example List last 100 ranked torrents in `Movies/x264/1080` category
    #   rarbg = RARBG::API.new
    #   rarbg.list(limit: 100, ranked: true, category: [44])
    #
    # @example List torrents with at least 50 seeders
    #   rarbg = RARBG::API.new
    #   rarbg.list(min_seeders: 50)
    #
    # @param params [Hash] A customizable set of parameters.
    #
    # @option params [Array<Integer>] :category Filter results by category.
    # @option params [Symbol] :format Format results.
    #   Accepted values: `:json`, `:json_extended`.
    #   Default: `:json`.
    # @option params [Integer] :limit Limit results number.
    #   Accepted values: `25`, `50`, `100`.
    #   Default: `25`.
    # @option params [Integer] :min_seeders Filter results by minimum seeders.
    # @option params [Integer] :min_leechers Filter results by minimum leechers.
    # @option params [Boolean] :ranked Include/exclude unranked torrents.
    #   Default: `true`.
    # @option params [Symbol] :sort Sort results.
    #   Accepted values: `:last`, `:seeders`, `:leechers`.
    #   Default: `:last`.
    #
    # @return [Array<Hash>] Torrents that match the specified parameters.
    #
    # @raise [ArgumentError] If `params` is not an `Hash`.
    # @raise [RARBG::APIError] If the request fails or the endpoint responds
    #   with an error.
    # @raise [Faraday::Error] On low-level connection errors (e.g. timeouts).
    def list(params = {})
      raise ArgumentError, 'Expected params hash' unless params.is_a?(Hash)

      params.update(mode: 'list', token: token?)
      call(params)
    end

    # Search torrents.
    #
    # @example Search by IMDb ID, sorted by leechers and in extended format.
    #   rarbg = RARBG::API.new
    #   rarbg.search(imdb: 'tt2488496', sort: :leechers, format: :json_extended)
    #
    # @example Search unranked torrents by string, with at least 2 seeders.
    #   rarbg = RARBG::API.new
    #   rarbg.search(string: 'Star Wars', ranked: false, min_seeders: 2)
    #
    # @param params [Hash] A customizable set of parameters.
    #
    # @option params [String] :string Search by string.
    # @option params [String] :imdb Search by IMDb id.
    # @option params [String] :tvdb Search by TVDB id.
    # @option params [String] :themoviedb Search by The Movie DB id.
    # @option params [Array<Integer>] :category Filter results by category.
    # @option params [Symbol] :format Format results.
    #   Accepted values: `:json`, `:json_extended`.
    #   Default: `:json`.
    # @option params [Integer] :limit Limit results number.
    #   Accepted values: `25`, `50`, `100`.
    #   Default: `25`.
    # @option params [Integer] :min_seeders Filter results by minimum seeders.
    # @option params [Integer] :min_leechers Filter results by minimum leechers.
    # @option params [Boolean] :ranked Include/exclude unranked torrents.
    #   Default: `true`.
    # @option params [Symbol] :sort Sort results.
    #   Accepted values: `:last`, `:seeders`, `:leechers`.
    #   Default: `:last`.
    #
    # @return [Array<Hash>] Torrents that match the specified parameters.
    #
    # @raise [ArgumentError] If `params` is not an `Hash`.
    # @raise [ArgumentError] If no search type param is passed
    #   (among `string`, `imdb`, `tvdb`, `themoviedb`).
    # @raise [RARBG::APIError] If the request fails or the endpoint responds
    #   with an error.
    # @raise [Faraday::Error] On low-level connection errors (e.g. timeouts).
    def search(params = {})
      raise ArgumentError, 'Expected params hash' unless params.is_a?(Hash)

      params.update(mode: 'search', token: token?)
      call(params)
    end

    # Generate the authentication token.
    #
    # @example Generate the token immediately after object instantiation.
    #   rarbg = RARBG::API.new
    #   rarbg.token!
    #
    # @return [String] The currently valid authentication token.
    def token!
      token?
    end

    private

    # Supported search parameters.
    SEARCH_KEYS = %w[string imdb tvdb themoviedb].freeze
    private_constant :SEARCH_KEYS

    # Endpoint error strings to return as no results.
    NO_RESULTS_ERRORS_REGEXPS = Regexp.union([
      /no results found/i,
      /can'?t find .+? in database/i
    ])
    private_constant :NO_RESULTS_ERRORS_REGEXPS

    # Wrap requests for error handling.
    def call(params)
      response = request(validate(params))

      return [] if response['error'] =~ NO_RESULTS_ERRORS_REGEXPS
      raise APIError, response['error'] if response.key?('error')

      response.fetch('torrent_results', [])
    end

    # Validate parameters.
    def validate(params)
      params = stringify(params)
      params = validate_search!(params) if params['mode'] == 'search'

      normalize.each_pair do |key, proc|
        params[key] = proc.call(params[key]) if params.key?(key)
      end

      params
    end

    # Convert symbol keys to string and remove nil values.
    def stringify(params)
      Hash[params.reject { |_k, v| v.nil? }.map { |k, v| [k.to_s, v] }]
    end

    # Validate search type parameters.
    def validate_search!(params)
      if (params.keys & SEARCH_KEYS).none?
        raise(ArgumentError,
              "One search parameter required among: #{SEARCH_KEYS.join(', ')}")
      end

      SEARCH_KEYS.each do |k|
        params["search_#{k}"] = params.delete(k) if params.key?(k)
      end

      params
    end

    # Convert ruby syntax to expected value format.
    def normalize
      {
        'category'    => (->(v) { v.join(';') }),
        'ranked'      => (->(v) { v == false ? 0 : 1 }),
        'search_imdb' => (->(v) { v.to_s[/^tt/] ? v.to_s : "tt#{v}" })
      }
    end

    # Return or renew auth token.
    def token?
      if token.nil? || time >= (token_time.to_f + TOKEN_EXPIRATION)
        response    = request(get_token: 'get_token')
        @token      = response.fetch('token')
        @token_time = time
      end

      token
    end

    # Perform API request.
    def request(params)
      rate_limit!(RATE_LIMIT)

      response      = connection.get(nil, params)
      @last_request = time

      return JSON.parse(response.body) if response.success?

      raise APIError, "#{response.reason_phrase} (#{response.status})"
    end

    # Rate-limit requests to comply with endpoint limits.
    def rate_limit!(seconds)
      sleep(0.1) until time >= (last_request.to_f + seconds)
    end

    # Monotonic clock for elapsed time calculations.
    def time
      Process.clock_gettime(Process::CLOCK_MONOTONIC)
    end
  end
end
